### ISSUE TITLE
Incorrect report of success when CLI commands actually fail

### ISSUE THREAD

**Issue by tlylt** [MarkBind/markbind#1863]
**Tell us about your environment**
Windows 10
* **MarkBind Version:**
3.1.1

**What did you do? Please include the actual source code causing the issue.**
When there is an error in site.json, or due to a mismatch of markbind version, the CLI commands output a warning but still run to "completion". This may not be ideal because for example in CI checks, it is not obvious that a (critical error) has occurred and because of that, the build step in CI may still be deemed successful (when it is not).

For example, when there is an error in the `site.json` file. e.g just remove a comma:
```
{
  "baseUrl": "/test-md",
  "titlePrefix": "",
```
to
```
{
  "baseUrl": "/test-md"
  "titlePrefix": "",
```

1. markbind serve output (which is ok since it actually reports an error)
![image](https://user-images.githubusercontent.com/41845017/160549312-8a62218d-f73e-4614-8907-7cd1a6538588.png)

2. markbind build output (no files being built but still says build success)
![image](https://user-images.githubusercontent.com/41845017/160549451-d68afbd5-025f-4268-8f6f-a3ba01b612be.png)

Additionally, if there is mismatch between markbind version, markbind build behaves similar to the above (which is not ideal), markbind serve also does not actually stop.

e.g. trying to serve the markbind site of cs3218 using the latest master branch markbind(there is a problem because the site uses .mbd files). The markbind serve is unable to serve any pages, but it runs to completion
![image](https://user-images.githubusercontent.com/41845017/160550003-3e1f2e57-0a25-4b9c-936e-0accebc9f5b2.png)
![image](https://user-images.githubusercontent.com/41845017/160550216-0393a8f4-6192-4835-8223-657f8dd1b58c.png)

**What did you expect to happen?**
I expect the command to fail so that it's failure can be picked up in CI.

**What actually happened? Please include the actual, raw output.**
See above, does not always fail and only show warnings.

**Additional context**
I tested build and serve of a docusaurus site and both will fail immediately if there is a problem with the `docusaurus.config.js`
![image](https://user-images.githubusercontent.com/41845017/160550691-87298c0e-b59d-4c39-b41c-2f80a86d1465.png)

This problem is related to the teammates devdocs migration. The issue is that given the fact that the build might fail, it should be included in the CI to ensure this is checked everytime a PR is merged.

With ref to this comment: https://github.com/TEAMMATES/teammates/pull/11680#issuecomment-1080854885, it may not be enough to use PR preview to resolve this

I tested in GitHub Action [here](https://github.com/tlylt/test-md/runs/5732354655?check_suite_focus=true) and if u see the logs, the build is failing because of the site.json file. But the action is still marked as successful.

Need some advice on
- is this a legitimate bug
- whether there is a solution to this that we can use to capture the error/warning

@MarkBind/active-devs

**Comment by kaixin-hc** [comment]
I think it would be preferred to throw an error, because checking a successful run is quite unlikely and it will be difficult to find the error quickly

**Comment by tlylt** [comment]
> I think it would be preferred to throw an error, because checking a successful run is quite unlikely and it will be difficult to find the error quickly

At the moment the only walkaround that I can think of is to check the existence of the `_site` folder after running markbind build. But I think this approach may not be 100% correct as the folder might(not sure) still be generated even though there's an error.

**Comment by jonahtanjz** [comment]
> But I think this approach may not be 100% correct as the folder might(not sure) still be generated even though there's an error.

Yes most of the errors during build will remove the `_site` folder except for the [`buildAssets`](https://github.com/MarkBind/markbind/blob/master/packages/core/src/Site/index.js#L908) step. This should not be a big problem as assets would not affect the built site too much in the case of teammates dev docs.

**Comment by jonahtanjz** [comment]
> Currently, the only exceptional case that I know of is when the MarkBind version is not compatible.

Seems like an issue with the `.mbd` file for this case. As for the teammates dev docs site, I think this should be fine as long as we inform them not to use `.mbd` files (doesn't seem like there are other breaking changes that will cause any incompatibility issue from v3.1.1 to v4)? This can just be a temporary solution from now till the release of v4.

**Comment by ang-zeyu** [comment]
(presumably, I dont have context too)

We've just haven't a well thought out error handling (I think since day 0 of markbind), and have been prioritsing usability in `markbind serve`.

E.g. we have this function https://github.com/MarkBind/markbind/blob/master/packages/core/src/html/elements.js#L7, which prefers presenting the error directly into the site (for visual appeal - presumably) rather than in the console.

The other more important reason would likely simply be to avoid having to restart `markbind serve` when there are development errors - something I think we should definitely aim to keep.

But I also agree we should fail-hard in the case of `markbind build`

---

labelling `d.hard` as it requires thinking through the entire application architecture and error handling we have currently. We might also want to attempt this only after typescript migration of most files, to prevent silly regressions.

Downgrading priority slightly as it has been the case since day 0, and there are very valid alternatives. I.e. check via `markbind serve` if everything works before deploying, as we have been doing. (which _should_ account for most errors)
